### DISCUSSION TITLE
Control input parameter consistently in multi-level multiapps

### DISCUSSION THREAD

**Discussion by dewenyushu** [idaholab/moose#19203] [Q&A Modules: General] [answered]
I would like to gather output data from a problem with different input parameters. The problem is designed as a multi-app problem, where one master app calling the full solve of one sub-app (master_app ->sub_app). The master-app and sub-app solves for different physics, but shares some input parameters that I'd like to perturb.

So what I am trying to do is to have a control app, which samples the input parameters and controls the master-app by using `MultiAppCommandLineControl`. However, I am not sure how to change the input parameters in the sub-app accordingly (e.g., for one sample, I'd like to keep the input parameters consistent between the master-app and the sub-app).

**Comment by dewenyushu** [comment]
@zachmprince would you mind providing some suggestions here?

**Comment by zachmprince** [comment, marked as answer]
This is how I would do this. So you have three files: stm.i (the stochastic tools driver), main.i (the main physics app), and sub.i (the sub physics app). Let's say you want to perturb the parameter `value` in a `BodyForce` kernel of a sub app.

sub.i:
```
[Kernels]
  [force]
    type = BodyForce
    variable = u
    value = 1 # Fake value
  []
[]
```

main.i
```
bf_value = 1 # Fake value, which is sent to the sub app

[MultiApps]
  [sub]
    ...
    cli_args = 'Kernels/force/value=${bf_value}'
  []
[]
```

stm.i:
```
[Controls]
  [cmdline]
    type = MultiAppCommandLineControl
    ...
    param_names = 'bf_value'
  []
[]
```

The stm app will control the parameter `bf_value` in the main app which will subsequently send it to the kernel parameter in the sub app.

**Reply by dewenyushu** [reply]
Thanks @zachmprince ! It is really nice that we can put `cli_args ` in the `MultiApps `. 

It seems to be the case though that we can only have one `cli_args` per sub-app. Any idea how to pass multiple parameters for one sub-app?

**Reply by zachmprince** [reply]
You can have multple values using `cli_args='Kernels/force1/value=${bf_value1};Kernels/force2/value=${bf_value2}`. Note the `;` separating the parameters and there can't be any spaces.

**Reply by dewenyushu** [reply]
Oh that's right - I was using spaces, which were recognized as one input per sub-app. Thanks Zach! This works great now!

**Reply by GiudGiud** [reply]
Let's document this trick with ; somewhere btw
